src/useSessionStorage/useSessionStorage.ts

```diff
@@ -16,27 +16,34 @@ declare global {
 
 type SetValue<T> = Dispatch<SetStateAction<T>>
 
-function useSessionStorage<T>(key: string, initialValue: T): [T, SetValue<T>] {
+function useSessionStorage<T>(
+  key: string,
+  initialValue: T | (() => T),
+): [T, SetValue<T>] {
   // Get from session storage then
   // parse stored json or return initialValue
   const readValue = useCallback((): T => {
+    const getInitialValue = () =>
+      initialValue instanceof Function ? initialValue() : initialValue
     // Prevent build error "window is undefined" but keep keep working
     if (typeof window === 'undefined') {
-      return initialValue
+      return getInitialValue()
     }
 
     try {
       const item = window.sessionStorage.getItem(key)
-      return item ? (parseJSON(item) as T) : initialValue
+      return item ? (parseJSON(item) as T) : getInitialValue()
     } catch (error) {
       console.warn(`Error reading sessionStorage key “${key}”:`, error)
-      return initialValue
+      return getInitialValue()
     }
   }, [initialValue, key])
 
   // State to store our value
   // Pass initial state function to useState so logic is only executed once
   const [storedValue, setStoredValue] = useState<T>(readValue)
+  // create a new empty object to keep track of the identity of the hook
+  const [source] = useState({})
```

Comment:
@juliencrn do you think it worthwhile to extract the creation of the empty object into a function so that it isn't created on each render?